File: project/.sdk/tm/php/feature/TestFeature.php
<?php
declare(strict_types=1);

// ProjectName SDK test feature

require_once __DIR__ . '/BaseFeature.php';
require_once __DIR__ . '/../utility/Param.php';

class ProjectNameTestFeature extends ProjectNameBaseFeature
{
    private mixed $client;
    private ?array $options;

    public function __construct()
    {
        parent::__construct();
        $this->version = '0.0.1';
        $this->name = 'test';
        $this->active = true;
        $this->client = null;
        $this->options = null;
    }

    public function init(ProjectNameContext $ctx, array $options): void
    {
        $this->client = $ctx->client;
        $this->options = $options;

        $entity_data = \Voxgig\Struct\Struct::getprop($options, 'entity');
        if (!is_array($entity_data)) {
            $entity_data = [];
        }

        $this->client->mode = 'test';

        // Ensure entity ids are correct.
        \Voxgig\Struct\Struct::walk($entity_data, function ($key, $val, $parent, $path) {
            if (count($path) === 2 && is_array($val) && $key !== null) {
                $val['id'] = $key;
            }
            return $val;
        });

        // Wrap in object so closure mutations persist (PHP arrays are value types).
        $entity = new \stdClass();
        $entity->data = $entity_data;

        $test_fetcher = function (ProjectNameContext $fctx, string $_fullurl, array $_fetchdef) use ($entity): array {
            $respond = function (int $status, mixed $data, ?array $extra = null): array {
                $out = [
                    'status' => $status,
                    'statusText' => 'OK',
                    'json' => function () use ($data) { return $data; },
                    'body' => 'not-used',
                ];
                if ($extra) {
                    foreach ($extra as $k => $v) {
                        $out[$k] = $v;
                    }
                }
                return [$out, null];
            };

            $op = $fctx->op;
            $entname = $op->entity;
            $entmap = is_array($entity->data[$entname] ?? null) ? $entity->data[$entname] : [];

            // PHP-portable equivalent of TS buildArgs+select: a flat-key
            // filter that matches by exact-equality on each provided key,
            // with alias fallback. Empty match matches all entries — load
            // with empty match returns the first fixture entry (or last
            // create), list returns all entries.
            $find_first = function (array $entmap, $match, $alias) {
                if (!is_array($match) || empty($match)) {
                    foreach ($entmap as $e) {
                        if (is_array($e)) return $e;
                    }
                    return null;
                }
                foreach ($entmap as $e) {
                    if (!is_array($e)) continue;
                    $ok = true;
                    foreach ($match as $k => $v) {
                        if ($v === null || $v === '__UNDEFINED__') continue;
                        $ev = $e[$k] ?? null;
                        if ($ev !== $v) {
                            // Try alias key if any
                            $ka = is_array($alias) ? ($alias[$k] ?? null) : null;
                            $aliased = ($ka !== null) ? ($e[$ka] ?? null) : null;
                            if ($aliased !== $v) {
                                $ok = false;
                                break;
                            }
                        }
                    }
                    if ($ok) return $e;
                }
                return null;
            };

            $find_all = function (array $entmap, $match, $alias) use ($find_first) {
                if (!is_array($match) || empty($match)) {
                    return array_values(array_filter($entmap, 'is_array'));
                }
                $out = [];
                foreach ($entmap as $e) {
                    if (!is_array($e)) continue;
                    $ok = true;
                    foreach ($match as $k => $v) {
                        if ($v === null || $v === '__UNDEFINED__') continue;
                        $ev = $e[$k] ?? null;
                        if ($ev !== $v) {
                            $ka = is_array($alias) ? ($alias[$k] ?? null) : null;
                            $aliased = ($ka !== null) ? ($e[$ka] ?? null) : null;
                            if ($aliased !== $v) { $ok = false; break; }
                        }
                    }
                    if ($ok) $out[] = $e;
                }
                return $out;
            };

            $alias = is_object($op) ? ($op->alias ?? null) : \Voxgig\Struct\Struct::getprop($op, 'alias');

            // For single-entity ops (load, remove) with an empty explicit
            // match, fall back to the id the entity client already knows from a
            // prior create/load (carried in $fctx->match / $fctx->data). This
            // mirrors the TS mock where param() resolves the id from that
            // accumulated state — e.g. `create()` then `remove([])` removes the
            // just-created entity, not an arbitrary fixture.
            $resolve_match = function ($explicit) use ($fctx) {
                if (is_array($explicit) && !empty($explicit)) return $explicit;
                foreach ([$fctx->match, $fctx->data] as $src) {
                    $arr = is_array($src) ? $src : (is_object($src) ? (array) $src : []);
                    if (isset($arr['id']) && $arr['id'] !== null && $arr['id'] !== '__UNDEFINED__') {
                        return ['id' => $arr['id']];
                    }
                }
                return [];
            };

            if ($op->name === 'load') {
                $ent = $find_first($entmap, $resolve_match($fctx->reqmatch), $alias);
                if ($ent === null) {
                    return $respond(404, null, ['statusText' => 'Not found']);
                }
                if (is_array($ent)) unset($ent['$KEY']);
                $out = \Voxgig\Struct\Struct::clone($ent);
                return $respond(200, $out);

            } elseif ($op->name === 'list') {
                $found = $find_all($entmap, $fctx->reqmatch, $alias);
                $cleaned = [];
                foreach ($found as $e) {
                    if (is_array($e)) unset($e['$KEY']);
                    $cleaned[] = $e;
                }
                $out = \Voxgig\Struct\Struct::clone($cleaned);
                return $respond(200, $out);

            } elseif ($op->name === 'update') {
                // Match the existing entity by id only (or its alias). reqdata
                // also contains the new field values, which would otherwise
                // cause find_first to filter out the entity we want to update.
                // When reqdata has no id, fall back to the id the entity
                // client carries from a prior create/load (in $fctx->match /
                // $fctx->data), mirroring the TS mock where param(ctx,'id')
                // resolves from accumulated state.
                $update_match = [];
                if (is_array($fctx->reqdata)) {
                    if (array_key_exists('id', $fctx->reqdata)) {
                        $update_match['id'] = $fctx->reqdata['id'];
                    }
                    $id_alias = is_array($alias) ? ($alias['id'] ?? null) : null;
                    if ($id_alias !== null && array_key_exists($id_alias, $fctx->reqdata)) {
                        $update_match[$id_alias] = $fctx->reqdata[$id_alias];
                    }
                }
                if (empty($update_match)) {
                    $update_match = $resolve_match([]);
                }
                $ent = $find_first($entmap, $update_match, $alias);
                if ($ent === null) {
                    return $respond(404, null, ['statusText' => 'Not found']);
                }
                if (is_array($fctx->reqdata)) {
                    foreach ($fctx->reqdata as $k => $v) {
                        $ent[$k] = $v;
                    }
                }
                $id = is_array($ent) ? ($ent['id'] ?? null) : null;
                if ($id !== null) {
                    $entmap[$id] = $ent;
                    $entity->data[$entname] = $entmap;
                }
                if (is_array($ent)) unset($ent['$KEY']);
                $out = \Voxgig\Struct\Struct::clone($ent);
                return $respond(200, $out);

            } elseif ($op->name === 'remove') {
                $ent = $find_first($entmap, $resolve_match($fctx->reqmatch), $alias);
                // Remove only the first matched entity. If nothing matches,
                // succeed as a no-op rather than erroring.
                $id = is_array($ent) ? ($ent['id'] ?? null) : null;
                if ($id !== null) {
                    unset($entmap[$id]);
                    $entity->data[$entname] = $entmap;
                }
                return $respond(200, null);

            } elseif ($op->name === 'create') {
                $id = ProjectNameParam::call($fctx, 'id');
                if ($id === null || $id === '__UNDEFINED__') {
                    $id = sprintf('%04x%04x%04x%04x',
                        random_int(0, 0xFFFF), random_int(0, 0xFFFF),
                        random_int(0, 0xFFFF), random_int(0, 0xFFFF));
                }

                $ent = is_array($fctx->reqdata) ? $fctx->reqdata : [];
                $ent['id'] = $id;
                $entmap[$id] = $ent;
                $entity->data[$entname] = $entmap;
                if (is_array($ent)) unset($ent['$KEY']);
                $out = \Voxgig\Struct\Struct::clone($ent);
                return $respond(200, $out);

            } else {
                return $respond(404, null, ['statusText' => 'Unknown operation']);
            }
        };

        $ctx->utility->fetcher = $test_fetcher;
    }

    /**
     * Build a structured `$AND` query from the request match/data dict,
     * matching the TS test feature's buildArgs. Mirrors ts/src/feature/test/TestFeature.ts:158-204.
     *
     * For each key in $args that is 'id' OR a required-param key on the
     * current operation point, emit a `$OR` clause matching the key (and
     * its alias, if any) against the supplied value.
     */
    public function buildArgs(ProjectNameContext $ctx, $op, $args): array
    {
        // If args is empty/missing, return an empty $AND so select() matches
        // every entry — the TS test feature relies on this for empty-match
        // load against fixture entries.
        $keys = is_array($args) ? \Voxgig\Struct\Struct::keysof($args) : [];
        if (empty($keys)) {
            return ['$AND' => []];
        }

        $opname = is_object($op) ? ($op->name ?? null) : (\Voxgig\Struct\Struct::getprop($op, 'name'));
        $entityName = null;
        if (isset($ctx->entity)) {
            $entityName = is_object($ctx->entity)
                ? ($ctx->entity->name ?? null)
                : (is_array($ctx->entity) ? ($ctx->entity['name'] ?? null) : null);
        }

        // Resolve required-param names from the op's last point. Defensive:
        // any missing piece falls back to "no required params".
        $reqd_names = [];
        if (is_string($opname) && is_string($entityName) && isset($ctx->config)) {
            $points = \Voxgig\Struct\Struct::getpath(
                ['entity', $entityName, 'op', $opname, 'points'],
                $ctx->config
            );
            $point = \Voxgig\Struct\Struct::getelem($points, -1);
            $params = is_array($point) ? ($point['args']['params'] ?? null) : null;
            if (is_array($params)) {
                foreach ($params as $p) {
                    if (is_array($p) && (($p['reqd'] ?? false) === true)) {
                        $n = $p['name'] ?? null;
                        if ($n !== null) {
                            $reqd_names[] = $n;
                        }
                    }
                }
            }
        }

        $alias = is_object($op) ? ($op->alias ?? null) : \Voxgig\Struct\Struct::getprop($op, 'alias');
        $qand = [];

        foreach ($keys as $k) {
            $is_id = ($k === 'id');
            $in_reqd = in_array($k, $reqd_names, true);
            if ($is_id || $in_reqd) {
                $v = ProjectNameParam::call($ctx, $k);
                $ka = \Voxgig\Struct\Struct::getprop($alias, $k);

                $qor = [[$k => $v]];
                if ($ka !== null) {
                    $qor[] = [$ka => $v];
                }

                $qand[] = ['$OR' => $qor];
            }
        }

        return ['$AND' => $qand];
    }
}

File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/TestFeature.swift
// In-memory mock transport for testing without a live server. Serves the
// per-entity fixture data supplied via the test feature options (`entity`) and
// answers CRUD requests the way a REST server would. An optional `net` block
// layers deterministic network-behaviour simulation over the mock.

import Foundation

private func testRespond(_ status: Int, _ data: Value, _ extra: VMap?) -> Value {
  let res = VMap()
  res.entries["status"] = .int(Int64(status))
  res.entries["statusText"] = .string("OK")
  let captured = data
  res.entries["json"] = .nat({ () -> Value in captured } as NativeCall0)
  res.entries["body"] = .string("not-used")
  if let extra = extra {
    for (k, v) in extra.entries { res.entries[k] = v }
  }
  return .map(res)
}

// For single-entity ops with an empty explicit match, fall back to the id the
// entity client already knows from a prior create/load.
private func testResolveMatch(_ ctx2: Context, _ explicitMatch: VMap) -> VMap {
  if explicitMatch.entries.count > 0 { return explicitMatch }
  for src in [ctx2.match, ctx2.data] {
    let v = gp(src, "id")
    if !isNil(v) && v != .string("__UNDEFINED__") {
      let m = VMap()
      m.entries["id"] = v
      return m
    }
  }
  return VMap()
}

private func testBuildArgs(_ ctx: Context, _ op: Operation, _ args: VMap?) -> Value {
  let opname = op.name

  let cfg: Value = ctx.config == nil ? .noval : .map(ctx.config!)
  let points = getpath(cfg, jtp("entity", ctx.entity!.getName(), "op", opname, "points"))
  let point = getelem(points, .int(-1))

  let paramsPath = getpath(point, jtp("args", "params"))
  let reqdParams = select(paramsPath, .map(vm(("reqd", .bool(true)))))
  let reqd = transform(reqdParams, jtp("`$EACH`", "", "`$KEY.name`"))

  let qand = VList()
  let q = VMap()
  q.entries["`$AND`"] = .list(qand)

  if let args = args {
    for key in keysof(.map(args)) {
      let isId = key == "id"
      let selected = select(reqd, .string(key))
      let isReqd = !isempty(selected)

      if isId || isReqd {
        let v = ctx.utility!.param(ctx, .string(key))
        let ka = gp(op.alias, key)

        let qor = VList()
        let qm1 = VMap()
        qm1.entries[key] = v
        qor.items.append(.map(qm1))
        if let kas = ka.asString {
          let qm2 = VMap()
          qm2.entries[kas] = v
          qor.items.append(.map(qm2))
        }

        let qorm = VMap()
        qorm.entries["`$OR`"] = .list(qor)
        qand.items.append(.map(qorm))
      }
    }
  }

  if let explain = ctx.ctrl.explain {
    let tm = VMap()
    tm.entries["query"] = .map(q)
    explain.entries["test"] = .map(tm)
  }

  return .map(q)
}

public final class TestFeature: BaseFeature {
  private var client: ProjectNameSDK?
  private var options: VMap?
  private var netcalls = 0

  public override init() {
    super.init()
    version = "0.0.1"
    name = "test"
    active = true
  }

  public override func initFeature(_ ctx: Context, _ options: VMap) {
    client = ctx.client
    self.options = options

    let entity = gp(options, "entity").asMap ?? VMap()

    client!.mode = "test"

    // Ensure entity ids are correct.
    _ = walk(.map(entity), { key, val, _, path in
      if path.count == 2, val.isMap, key.asString != nil {
        val.asMap!.entries["id"] = .string(strkey(key))
      }
      return val
    }, nil)

    let testFetcher: FetcherFunc = { ctx2, _fullurl, _fetchdef in
      let op = ctx2.op!
      let entmap = gp(entity, op.entity).asMap ?? VMap()

      if op.name == "load" {
        let args = testBuildArgs(ctx2, op, testResolveMatch(ctx2, ctx2.reqmatch))
        let found = select(.map(entmap), args)
        let ent = getelem(found, .int(0))
        if isNil(ent) {
          let extra = VMap()
          extra.entries["statusText"] = .string("Not found")
          return testRespond(404, .noval, extra)
        }
        delprop(ent, .string("$KEY"))
        return testRespond(200, clone(ent), nil)
      } else if op.name == "list" {
        let args = testBuildArgs(ctx2, op, ctx2.reqmatch)
        let found = select(.map(entmap), args)
        if isNil(found) {
          let extra = VMap()
          extra.entries["statusText"] = .string("Not found")
          return testRespond(404, .noval, extra)
        }
        if let fl = found.asList {
          for item in fl.items { delprop(item, .string("$KEY")) }
        }
        return testRespond(200, clone(found), nil)
      } else if op.name == "update" {
        var updateMatch = VMap()
        if let idv = ctx2.reqdata.entries["id"] {
          updateMatch.entries["id"] = idv
        }
        if let alias = op.alias, let aliasId = gp(alias, "id").asString,
          let av = ctx2.reqdata.entries[aliasId] {
          updateMatch.entries[aliasId] = av
        }
        if updateMatch.entries.count == 0 {
          updateMatch = testResolveMatch(ctx2, VMap())
        }
        let args = testBuildArgs(ctx2, op, updateMatch)
        let found = select(.map(entmap), args)
        var ent = getelem(found, .int(0))
        if isNil(ent) {
          for (_, e) in entmap.entries where e.isMap { ent = e; break }
        }
        if isNil(ent) {
          let extra = VMap()
          extra.entries["statusText"] = .string("Not found")
          return testRespond(404, .noval, extra)
        }
        if let entm = ent.asMap {
          for (k, v) in ctx2.reqdata.entries { entm.entries[k] = v }
        }
        delprop(ent, .string("$KEY"))
        return testRespond(200, clone(ent), nil)
      } else if op.name == "remove" {
        let args = testBuildArgs(ctx2, op, testResolveMatch(ctx2, ctx2.reqmatch))
        let found = select(.map(entmap), args)
        let ent = getelem(found, .int(0))
        if let entm2 = ent.asMap {
          let id = gp(entm2, "id")
          delprop(.map(entmap), id)
        }
        return testRespond(200, .noval, nil)
      } else if op.name == "create" {
        _ = testBuildArgs(ctx2, op, ctx2.reqdata)
        var id = ctx2.utility!.param(ctx2, .string("id"))
        if isNil(id) {
          id = .string(String(
            format: "%04x%04x%04x%04x",
            Int.random(in: 0..<0x10000), Int.random(in: 0..<0x10000),
            Int.random(in: 0..<0x10000), Int.random(in: 0..<0x10000)))
        }
        let ent = clone(.map(ctx2.reqdata))
        if let entm = ent.asMap {
          entm.entries["id"] = id
          if let idStr = id.asString { entmap.entries[idStr] = .map(entm) }
          delprop(.map(entm), .string("$KEY"))
          return testRespond(200, clone(.map(entm)), nil)
        }
        return testRespond(200, ent, nil)
      }

      let extra = VMap()
      extra.entries["statusText"] = .string("Unknown operation")
      return testRespond(404, .noval, extra)
    }

    // Optional network behaviour simulation over the mock transport.
    let net = gp(options, "net").asMap
    ctx.utility!.fetcher = net == nil ? testFetcher : makeNetsim(net!, testFetcher)
  }

  private func makeNetsim(_ net: VMap, _ inner: @escaping FetcherFunc) -> FetcherFunc {
    netcalls = 0

    func pickLatency() -> Int {
      let l = gp(net, "latency")
      if isNil(l) { return 0 }
      if let lm = l.asMap {
        let mn = foptInt(lm, "min", 0)
        let mx = foptInt(lm, "max", mn)
        if mx <= mn { return mn }
        return mn + ((mx - mn) >> 1)
      }
      let fixedMs = foptInt(net, "latency", 0)
      return fixedMs < 0 ? 0 : fixedMs
    }

    func sleepMs(_ ms: Int) {
      if ms <= 0 { return }
      foptSleep(net)(ms)
    }

    return { [weak self] ctx, url, fetchdef in
      guard let self = self else { return try inner(ctx, url, fetchdef) }
      self.netcalls += 1
      let call = self.netcalls

      if foptBool(net, "offline", false) {
        sleepMs(pickLatency())
        throw ctx.makeError("netsim_offline", "Simulated network offline (URL was: \"\(url)\")")
      }
      if call <= foptInt(net, "errorTimes", 0) {
        sleepMs(pickLatency())
        throw ctx.makeError("netsim_conn", "Simulated connection error (call \(call))")
      }
      if call <= foptInt(net, "failTimes", 0) {
        sleepMs(pickLatency())
        let status = foptInt(net, "failStatus", 503)
        let r = VMap()
        r.entries["status"] = .int(Int64(status))
        r.entries["statusText"] = .string("Simulated Failure")
        r.entries["body"] = .string("not-used")
        r.entries["json"] = .nat({ () -> Value in .noval } as NativeCall0)
        r.entries["headers"] = .map(VMap())
        return .map(r)
      }
      sleepMs(pickLatency())
      return try inner(ctx, url, fetchdef)
    }
  }
}
